### PR TITLE
Giza utest content extratests

### PR DESCRIPTION
**Pull request by dobertRowneySr**
Added tests in to account for the new content/storage fixes added in PR #2907.
Tests list is at #2983 together with issues found (and now fixed)

### CODE REVIEW

**Comment by mnaamani** [comment]
just a top @ignazio-bovo you should run `yarn cargo-checks` to run the linter when updating rust code. There is a minor linter warning in the joystream-node build.

```
error: use of `or` followed by a function call
    --> runtime-modules/storage/src/lib.rs:2881:18
     |
2881 |                   .or(upload_params
     |  __________________^
2882 | |                     .as_ref()
2883 | |                     .map(|p| p.deletion_prize_source_account_id.clone()));
     | |_________________________________________________________________________^
     |
     = note: `-D clippy::or-fun-call` implied by `-D warnings`
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#or_fun_call
help: try this
     |
2881 |                 .or_else(|| upload_params
2882 |                     .as_ref()
2883 |                     .map(|p| p.deletion_prize_source_account_id.clone()));
     |

error: aborting due to previous error

error: could not compile `pallet-storage`
```

File: runtime-modules/storage/src/tests/mod.rs [hunk 1]
```diff
@@ -10,6 +10,7 @@ use frame_system::RawOrigin;
 use sp_std::collections::btree_map::BTreeMap;
 use sp_std::collections::btree_set::BTreeSet;
 use sp_std::iter::{repeat, FromIterator};
+use std::convert::TryInto;
```
Comment (shamil-gadelshin) [inline runtime-modules/storage/src/tests/mod.rs]: Try using `sp_std` instead. Parity could override something in the future.

File: runtime-modules/storage/src/tests/fixtures.rs [hunk 2]
```diff
@@ -1179,25 +1245,62 @@ impl CreateDynamicBagWithObjectsFixture {
...
+                    let enough_size =
+                        bucket.voucher.size_limit >= total_size_required + bucket.voucher.size_used;
+                    let enough_objects = bucket.voucher.objects_limit
+                        >= total_objects_required + bucket.voucher.objects_used;
+                    enough_size && enough_objects && bucket.accepting_new_bags
+                }));
+            }
+            Err(err) => {
+                assert_eq!(balance_pre, balance_post);
+                if let DispatchError::Module {
```
Comment (shamil-gadelshin) [inline runtime-modules/storage/src/tests/fixtures.rs]: Can't this error check be simplified? Like other error checks in `tests/mod.rs`?

Comment (shamil-gadelshin) [reply runtime-modules/storage/src/tests/fixtures.rs]: Why do you expect the only error type here?

Comment (shamil-gadelshin) [reply runtime-modules/storage/src/tests/fixtures.rs]: Haven't you already checked the expected and actual results earlier?

File: runtime-modules/storage/src/tests/fixtures.rs [hunk 3]
```diff
@@ -1179,25 +1245,62 @@ impl CreateDynamicBagWithObjectsFixture {
...
 
-        if actual_result.is_ok() {
-            let bag_id: BagId<Test> = self.bag_id.clone().into();
-            assert!(<crate::Bags<Test>>::contains_key(&bag_id));
+        match actual_result {
+            Ok(()) => {
+                assert!(<crate::Bags<Test>>::contains_key(&bag_id));
+
+                let bag = crate::Bags::<Test>::get(&bag_id);
+                assert_eq!(
```
Comment (shamil-gadelshin) [inline runtime-modules/storage/src/tests/fixtures.rs:1272]: Is there any check for the data fee?

File: runtime-modules/content/src/tests/fixtures.rs [hunk 4]
```diff
@@ -0,0 +1,932 @@
...
+        let beg_obj_id = storage::NextDataObjectId::<Test>::get();
+        let actual_result =
+            Content::create_channel(origin, self.actor.clone(), self.params.clone());
+        let end_obj_id = storage::NextDataObjectId::<Test>::get();
+
+        assert_eq!(actual_result, expected_result);
+
+        let balance_post = Balances::usable_balance(self.sender);
+
+        match actual_result {
```
Comment (shamil-gadelshin) [inline runtime-modules/content/src/tests/fixtures.rs]: Could be simplified: if actual_result.is_ok()

File: runtime-modules/content/src/tests/fixtures.rs [hunk 5]
```diff
@@ -0,0 +1,932 @@
...
+            self.params.clone(),
+        );
+
+        let balance_post = Balances::usable_balance(self.sender);
+        let end_obj_id = storage::NextDataObjectId::<Test>::get();
+
+        assert_eq!(actual_result, expected_result);
+
+        match actual_result {
+            Ok(()) => {
```
Comment (shamil-gadelshin) [inline runtime-modules/content/src/tests/fixtures.rs]: Please, simplify such code.

**Review by shamil-gadelshin — CHANGES_REQUESTED** [review]
Good work! Most use cases are covered now.

There are several minor issues.

File: runtime-modules/storage/src/tests/mod.rs (hunk 1 above)
Comment (ignazio-bovo) [reply runtime-modules/storage/src/tests/mod.rs:13]: fixed in commit`849a44c5fc`

File: runtime-modules/storage/src/tests/fixtures.rs (hunk 3 above)
Comment (dobertRowneySr) [reply runtime-modules/storage/src/tests/fixtures.rs:1272]: `create_dynamic_bag_with_object_constraints` only transfer the bag deletion prize to the storage treasury account. Slashing `upload_data_fee`s & transferring `object.deletion_prize_fee`s is left to `upload_data_object`, which is called right after in the `create_channel` extrinsic

File: runtime-modules/content/src/tests/fixtures.rs (hunk 4 above)
Comment (ignazio-bovo) [reply runtime-modules/content/src/tests/fixtures.rs:93]: I have added a hopefully cleaner solution

File: runtime-modules/storage/src/tests/fixtures.rs (hunk 2 above)
Comment (dobertRowneySr) [reply runtime-modules/storage/src/tests/fixtures.rs]: Tried to simplify the pattern matching in 36524997d136120f6476fd554f22729e85b4962a

File: runtime-modules/content/src/tests/fixtures.rs (hunk 5 above)
Comment (dobertRowneySr) [reply runtime-modules/content/src/tests/fixtures.rs]: Tried to simplify in c80eb1a1a5055ba3c033cecf6ca90fe9decf16db

File: runtime-modules/content/src/tests/fixtures.rs (hunk 4 above)
Comment (dobertRowneySr) [reply runtime-modules/content/src/tests/fixtures.rs]: fixed in ebd78029211e0b9911592cbf28c406e8f9727ff4

File: runtime-modules/storage/src/tests/mod.rs (hunk 1 above)
Comment (dobertRowneySr) [reply runtime-modules/storage/src/tests/mod.rs]: fixed in 849a44c5fcf724896745b12c2ece7b5795e6ad32

File: runtime-modules/storage/src/tests/fixtures.rs [hunk 6]
```diff
@@ -1179,25 +1245,55 @@ impl CreateDynamicBagWithObjectsFixture {
...
+                    let enough_size =
+                        bucket.voucher.size_limit >= total_size_required + bucket.voucher.size_used;
+                    let enough_objects = bucket.voucher.objects_limit
+                        >= total_objects_required + bucket.voucher.objects_used;
+                    enough_size && enough_objects && bucket.accepting_new_bags
+                }));
+            }
+            Err(err) => {
+                assert_eq!(balance_pre, balance_post);
+                if into_str(err) != "DynamicBagExists" {
```
Comment (shamil-gadelshin) [inline runtime-modules/storage/src/tests/fixtures.rs:1293]: Why don't you use the Error object itself like in the calling function? 

```
CreateDynamicBagWithObjectsFixture::default()
            .with_objects(objects)
            .call_and_assert(Err(Error::<Test>::ZeroObjectSize.into()));
```

You're using the string representation that will result in a test fail on the error change.

**Review by shamil-gadelshin — CHANGES_REQUESTED** [review]
Almost there.

Please, fix string error messages.

File: runtime-modules/storage/src/tests/fixtures.rs (hunk 6 above)
Comment (dobertRowneySr) [reply runtime-modules/storage/src/tests/fixtures.rs:1293]: Hope I understood the problem correctly. Fixed in 09f58c963de1fc083dcc5f744d5c17016c8a3eb5

File: runtime-modules/content/src/permissions/mod.rs [hunk 7]
```diff
@@ -261,13 +261,16 @@ pub fn ensure_actor_authorized_to_update_channel_assets<T: Trait>(
...
-            // ensure valid member
+            // ensure member account and origin correspondence
             ensure_member_auth_success::<T>(sender, member_id)?;
             // ensure member is channel owner
             ensure_member_is_channel_owner::<T>(&channel.owner, member_id)?;
             Ok(())
         }
         ContentActor::Collaborator(member_id) => {
+            // ensure member account and origin correspondence
+            ensure_member_auth_success::<T>(sender, member_id)?;
```
Comment (mnaamani) [inline runtime-modules/content/src/permissions/mod.rs:272]: Its good to see while writing the tests you were able to catch this security bug.

File: runtime-modules/content/src/tests/mock.rs [hunk 8]
```diff
@@ -443,280 +466,122 @@ pub fn run_to_block(n: u64) {
...
-            meta: None,
-            reward_account: None,
-            collaborators: BTreeSet::new(),
-        },
-        Ok(()),
-    );
-}
+pub type Balances = balances::Module<Test>;
+
+// pub fn delete_channel_mock(
```
Comment (mnaamani) [inline runtime-modules/content/src/tests/mock.rs]: If this code is obsolete delete it instead of commenting it out

Comment (dobertRowneySr) [reply runtime-modules/content/src/tests/mock.rs]: Fixed in 800d054c47349329e12a6b4e0d00ba14747f781e

**Comment by mnaamani** [comment]
I tried to merge this into giza_staging to run the tests there but there are some issues to fix:
```
error[E0425]: cannot find value `FIRST_MEMBER_ORIGIN` in this scope
  --> runtime-modules/content/src/tests/migration.rs:30:28
   |
30 |             Origin::signed(FIRST_MEMBER_ORIGIN),
   |                            ^^^^^^^^^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find value `FIRST_MEMBER_ID` in this scope
  --> runtime-modules/content/src/tests/migration.rs:31:34
   |
31 |             ContentActor::Member(FIRST_MEMBER_ID),
   |                                  ^^^^^^^^^^^^^^^ help: a constant with a similar name exists: `DEFAULT_MEMBER_ID`
   | 
  ::: runtime-modules/content/src/tests/mock.rs:37:1
   |
37 | pub const DEFAULT_MEMBER_ID: MemberId = 201;
   | -------------------------------------------- similarly named constant `DEFAULT_MEMBER_ID` defined here

error[E0425]: cannot find value `FIRST_MEMBER_ORIGIN` in this scope
  --> runtime-modules/content/src/tests/migration.rs:44:28
   |
44 |             Origin::signed(FIRST_MEMBER_ORIGIN),
   |                            ^^^^^^^^^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find value `FIRST_MEMBER_ID` in this scope
  --> runtime-modules/content/src/tests/migration.rs:45:34
   |
45 |             ContentActor::Member(FIRST_MEMBER_ID),
   |                                  ^^^^^^^^^^^^^^^ help: a constant with a similar name exists: `DEFAULT_MEMBER_ID`
   | 
  ::: runtime-modules/content/src/tests/mock.rs:37:1
   |
37 | pub const DEFAULT_MEMBER_ID: MemberId = 201;
   | -------------------------------------------- similarly named constant `DEFAULT_MEMBER_ID` defined here

error[E0425]: cannot find value `FIRST_MEMBER_ORIGIN` in this scope
  --> runtime-modules/content/src/tests/migration.rs:53:28
   |
53 |             Origin::signed(FIRST_MEMBER_ORIGIN),
   |                            ^^^^^^^^^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find value `FIRST_MEMBER_ID` in this scope
  --> runtime-modules/content/src/tests/migration.rs:54:34
   |
54 |             ContentActor::Member(FIRST_MEMBER_ID),
   |                                  ^^^^^^^^^^^^^^^ help: a constant with a similar name exists: `DEFAULT_MEMBER_ID`
   | 
  ::: runtime-modules/content/src/tests/mock.rs:37:1
   |
37 | pub const DEFAULT_MEMBER_ID: MemberId = 201;
   | -------------------------------------------- similarly named constant `DEFAULT_MEMBER_ID` defined here

error[E0425]: cannot find value `FIRST_MEMBER_ORIGIN` in this scope
  --> runtime-modules/content/src/tests/migration.rs:68:28
   |
68 |             Origin::signed(FIRST_MEMBER_ORIGIN),
   |                            ^^^^^^^^^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find value `FIRST_MEMBER_ID` in this scope
  --> runtime-modules/content/src/tests/migration.rs:69:34
   |
69 |             ContentActor::Member(FIRST_MEMBER_ID),
   |                                  ^^^^^^^^^^^^^^^ help: a constant with a similar name exists: `DEFAULT_MEMBER_ID`
   | 
  ::: runtime-modules/content/src/tests/mock.rs:37:1
   |
37 | pub const DEFAULT_MEMBER_ID: MemberId = 201;
   | -------------------------------------------- similarly named constant `DEFAULT_MEMBER_ID` defined here

error[E0425]: cannot find value `FIRST_MEMBER_ORIGIN` in this scope
  --> runtime-modules/content/src/tests/migration.rs:82:28
   |
82 |             Origin::signed(FIRST_MEMBER_ORIGIN),
   |                            ^^^^^^^^^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find value `FIRST_MEMBER_ID` in this scope
  --> runtime-modules/content/src/tests/migration.rs:83:34
   |
83 |             ContentActor::Member(FIRST_MEMBER_ID),
   |                                  ^^^^^^^^^^^^^^^ help: a constant with a similar name exists: `DEFAULT_MEMBER_ID`
   | 
  ::: runtime-modules/content/src/tests/mock.rs:37:1
   |
37 | pub const DEFAULT_MEMBER_ID: MemberId = 201;
   | -------------------------------------------- similarly named constant `DEFAULT_MEMBER_ID` defined here

error[E0425]: cannot find value `FIRST_MEMBER_ORIGIN` in this scope
  --> runtime-modules/content/src/tests/migration.rs:93:28
   |
93 |             Origin::signed(FIRST_MEMBER_ORIGIN),
   |                            ^^^^^^^^^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find value `FIRST_MEMBER_ID` in this scope
  --> runtime-modules/content/src/tests/migration.rs:94:34
   |
94 |             ContentActor::Member(FIRST_MEMBER_ID),
   |                                  ^^^^^^^^^^^^^^^ help: a constant with a similar name exists: `DEFAULT_MEMBER_ID`
   | 
  ::: runtime-modules/content/src/tests/mock.rs:37:1
   |
37 | pub const DEFAULT_MEMBER_ID: MemberId = 201;
   | -------------------------------------------- similarly named constant `DEFAULT_MEMBER_ID` defined here

error[E0425]: cannot find value `FIRST_MEMBER_ORIGIN` in this scope
   --> runtime-modules/content/src/tests/migration.rs:104:28
    |
104 |             Origin::signed(FIRST_MEMBER_ORIGIN),
    |                            ^^^^^^^^^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find value `FIRST_MEMBER_ID` in this scope
   --> runtime-modules/content/src/tests/migration.rs:105:34
    |
105 |             ContentActor::Member(FIRST_MEMBER_ID),
    |                                  ^^^^^^^^^^^^^^^ help: a constant with a similar name exists: `DEFAULT_MEMBER_ID`
    | 
   ::: runtime-modules/content/src/tests/mock.rs:37:1
    |
37  | pub const DEFAULT_MEMBER_ID: MemberId = 201;
    | -------------------------------------------- similarly named constant `DEFAULT_MEMBER_ID` defined here

error[E0425]: cannot find value `FIRST_MEMBER_ORIGIN` in this scope
   --> runtime-modules/content/src/tests/migration.rs:113:28
    |
113 |             Origin::signed(FIRST_MEMBER_ORIGIN),
    |                            ^^^^^^^^^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find value `FIRST_MEMBER_ID` in this scope
   --> runtime-modules/content/src/tests/migration.rs:114:34
    |
114 |             ContentActor::Member(FIRST_MEMBER_ID),
    |                                  ^^^^^^^^^^^^^^^ help: a constant with a similar name exists: `DEFAULT_MEMBER_ID`
    | 
   ::: runtime-modules/content/src/tests/mock.rs:37:1
    |
37  | pub const DEFAULT_MEMBER_ID: MemberId = 201;
    | -------------------------------------------- similarly named constant `DEFAULT_MEMBER_ID` defined here

error[E0425]: cannot find value `FIRST_MEMBER_ORIGIN` in this scope
   --> runtime-modules/content/src/tests/migration.rs:124:10
    |
124 |         &FIRST_MEMBER_ORIGIN,
    |          ^^^^^^^^^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find function `create_channel_mock` in this scope
   --> runtime-modules/content/src/tests/migration.rs:130:9
    |
130 |         create_channel_mock(
    |         ^^^^^^^^^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find value `FIRST_MEMBER_ORIGIN` in this scope
   --> runtime-modules/content/src/tests/migration.rs:131:13
    |
131 |             FIRST_MEMBER_ORIGIN,
    |             ^^^^^^^^^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find value `FIRST_MEMBER_ID` in this scope
   --> runtime-modules/content/src/tests/migration.rs:132:34
    |
132 |             ContentActor::Member(FIRST_MEMBER_ID),
    |                                  ^^^^^^^^^^^^^^^ help: a constant with a similar name exists: `DEFAULT_MEMBER_ID`
    | 
   ::: runtime-modules/content/src/tests/mock.rs:37:1
    |
37  | pub const DEFAULT_MEMBER_ID: MemberId = 201;
    | -------------------------------------------- similarly named constant `DEFAULT_MEMBER_ID` defined here

error[E0425]: cannot find function `create_video_mock` in this scope
   --> runtime-modules/content/src/tests/migration.rs:150:9
    |
150 |         create_video_mock(
    |         ^^^^^^^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find value `FIRST_MEMBER_ORIGIN` in this scope
   --> runtime-modules/content/src/tests/migration.rs:151:13
    |
151 |             FIRST_MEMBER_ORIGIN,
    |             ^^^^^^^^^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find value `FIRST_MEMBER_ID` in this scope
   --> runtime-modules/content/src/tests/migration.rs:152:34
    |
152 |             ContentActor::Member(FIRST_MEMBER_ID),
    |                                  ^^^^^^^^^^^^^^^ help: a constant with a similar name exists: `DEFAULT_MEMBER_ID`
    | 
   ::: runtime-modules/content/src/tests/mock.rs:37:1
    |
37  | pub const DEFAULT_MEMBER_ID: MemberId = 201;
    | -------------------------------------------- similarly named constant `DEFAULT_MEMBER_ID` defined here

error[E0599]: no function or associated item named `ensure_channel_exists` found for struct `Module<mock::Test>` in the current scope
    --> runtime-modules/content/src/tests/fixtures.rs:94:33
     |
94   |               assert_ok!(Content::ensure_channel_exists(&channel_id));
     |                                   ^^^^^^^^^^^^^^^^^^^^^
     |                                   |
     |                                   function or associated item not found in `Module<mock::Test>`
     |                                   help: there is an associated function with a similar name: `ensure_channel_validity`
     | 
    ::: runtime-modules/content/src/lib.rs:551:1
     |
551  | / decl_module! {
552  | |     pub struct Module<T: Trait> for enum Call where origin: T::Origin {
553  | |         /// Predefined errors
554  | |         type Error = Error<T>;
...    |
1472 | |     }
1473 | | }
     | |_- function or associated item `ensure_channel_exists` not found for this

error[E0599]: no function or associated item named `ensure_video_exists` found for struct `Module<mock::Test>` in the current scope
    --> runtime-modules/content/src/tests/fixtures.rs:217:33
     |
217  |               assert_ok!(Content::ensure_video_exists(&video_id));
     |                                   ^^^^^^^^^^^^^^^^^^^
     |                                   |
     |                                   function or associated item not found in `Module<mock::Test>`
     |                                   help: there is an associated function with a similar name: `ensure_video_validity`
     | 
    ::: runtime-modules/content/src/lib.rs:551:1
     |
551  | / decl_module! {
552  | |     pub struct Module<T: Trait> for enum Call where origin: T::Origin {
553  | |         /// Predefined errors
554  | |         type Error = Error<T>;
...    |
1472 | |     }
1473 | | }
     | |_- function or associated item `ensure_video_exists` not found for this
```

I think it would be better if you could target this PR to giza_staging and resolve them at the same time.

**Review by shamil-gadelshin — APPROVED** [review]
LGTM

**Pull request merged**